### PR TITLE
Add support for reservation affinity, dataproc oss metric collection and node-group configs, also support 'SPOT' under preemptibility

### PR DESCRIPTION
**Pull request by duqDarren**
fixes https://github.com/hashicorp/terraform-provider-google/issues/12848

Adds the support of creating cluster config with resource affinity, dataproc oss metric collection,  node-group configs, and support 'SPOT' under preemptibility. 

If this PR is for Terraform, I acknowledge that I have:

- [ X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

Release Note Template for Downstream PRs (will be copied)

```release-note:enhancement
dataproc: added support for `reservation_affinity` in `google_dataproc_cluster`
```
```release-note:enhancement
dataproc: added support for `node_group_affinity` in `google_dataproc_cluster`
```
```release-note:enhancement
dataproc: added support for `dataproc_metric_config` in `google_dataproc_cluster`
```
```release-note:enhancement
dataproc: added support for `SPOT` option for `preemptibility` in `google_dataproc_cluster`
```

### CODE REVIEW

**Comment by duqDarren** [comment]
Hello @roaks3 , it seems like the failed tests cause by the system error, could you help to interpret these failed tests?

**Comment by roaks3** [comment]
@duqDarren Two checklist items before this is ready for review:
1. Could you please fix the release note per https://github.com/GoogleCloudPlatform/magic-modules/pull/6889#issuecomment-1333878233. It appears you're missing a type.
2. Could you fix the following test failure? It appears to be breaking the provider itself, and producing the large number of test failures we're seeing:
```
--- FAIL: TestProvider (0.08s)
    provider_test.go:434: err: 1 error occurred:
            * resource google_dataproc_cluster: values: Default is not valid for lists or sets
        
FAIL
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta 22.516s
?       github.com/hashicorp/terraform-provider-google-beta/version [no test files]
FAIL
```

**Comment by duqDarren** [comment]
Thanks Ryan, I removed the Default entry, and now the tests all passing.

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (roaks3) [inline mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:663]: We typically don't need to include unspecified values like `TYPE_UNSPECIFIED`. Do you know if it is necessary here for some reason?

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (roaks3) [inline mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:664]: Does this behave as expected when a value is not provided? It looks like the wrong type for this field since it is a string. I would consider not including a default here unless it is needed by the API.

**Review by roaks3 — CHANGES_REQUESTED** [review]
I had two small comments, but I think there are two larger items that we need before approval:
1. Could you add this new field to one of our tests to verify that it is working as intended? (ie. `mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb`)
2. Could you add this new field to the docs (maybe including it in an example as well)? (ie. `mmv1/third_party/terraform/website/docs/r/dataproc_cluster.html.markdown`)

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:663]: Hi Ryan, from out proto, it does has this field. http://shortn/_BYbuhExsdX. 
I also followed similar example at line 683, which includes PREEMPTIBILITY_UNSPECIFIED. 

Let me know if you have concerns on this, i can remove it if you still think its not expected.

**Comment by duqDarren** [comment]
>

**Comment by duqDarren** [comment]
> 

Added.

**Comment by duqDarren** [comment]
> I had two small comments, but I think there are two larger items that we need before approval:
> 
> 1. Could you add this new field to one of our tests to verify that it is working as intended? (ie. `mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb`)
> 2. Could you add this new field to the docs (maybe including it in an example as well)? (ie. `mmv1/third_party/terraform/website/docs/r/dataproc_cluster.html.markdown`)

Added

**Comment by roaks3** [comment]
Hey @duqDarren I just wanted to double check after I saw the activity in here related to the tests: is this ready for a final review?

**Comment by duqDarren** [comment]
> Hey @duqDarren I just wanted to double check after I saw the activity in here related to the tests: is this ready for a final review?

Hi @roaks3, yes, the PR now is ready for a final review. All checks & tests passed.

**Review by roaks3 — CHANGES_REQUESTED** [review]
I took a first pass at reviewing and overall this looks good. The big thing to resolve is the schemas, and then I would consider the incomplete tests to be a blocker.

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb]: Thanks Ryan pointing out this. I initially get confused to follow the gcloud cli or the API, since i saw the subnet_uri and network_uri followed the gcloud format, so i chose follow the gcloud cli as well. 

Since derive from API is prefered, i updated the schema based on the API.

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb]: Removed

File: mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb:1402]: i removed reservation and follow the API to add the new entry.

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb]: Done. changed to List

File: mmv1/third_party/terraform/website/docs/r/dataproc_cluster.html.markdown
Comment (duqDarren) [reply mmv1/third_party/terraform/website/docs/r/dataproc_cluster.html.markdown:623]: Sorry for the confusion, i guess my PR missed this part. 
I added this to the Preemptibility entry. This SPOT option supported by API, but missed the TF(https://cloud.google.com/dataproc/docs/reference/rest/v1/ClusterConfig#preemptibility)

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (roaks3) [inline mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:760]: What are your thoughts on adding a test for `SPOT`? I think since it appears to be a new option here, it would be good to confirm it works.

Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:760]: Added.

File: mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/resources/resource_dataproc_cluster.go.erb:672]: yes, i think we should match it with the API.

File: mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb
Comment (duqDarren) [reply mmv1/third_party/terraform/tests/resource_dataproc_cluster_test.go.erb:386]: Hi Ryan, this NodeGroupAffinity is actually a dependency on other GCE resources. We just want to test if this property was provided to the backend, so i think catch the right error msg should be able to verify this new field.

**Comment by cjac** [comment]
Please include mention of dataproc oss metric collection in the title for this PR and I will close #6788 a dup

**Comment by cjac** [comment]
@duqDarren can you please merge with master?

**Comment by cjac** [comment]
Thanks!  I merged and built it today, so my smoke test is pretty much done.  I would like to try the provider out.  Do you have a test environment set up where you can launch dataproc from a terraform file?

**Comment by roaks3** [comment]
@duqDarren I slightly tweaked the release notes for consistency and broke them into separate items

**Review by roaks3 — APPROVED** [review]

**Comment by melinath** [comment]
As a heads up, this will need to be reverted because it introduces some GA-only fields (which we do not allow in the provider). Additionally, one of those fields has a crashing bug. Please feel free to open a new PR to re-introduce the fields to both the GA and beta providers - ideally as separate PRs for each field.

**Pull request merged**
